### ISSUE TITLE
Add several actions based on commit history

### ISSUE THREAD

**Issue by jbsarrodie** [archi-contribs/archi-modelrepository-plugin#16]
Some potential actions:
* Open the version of the model at a specified commit: this would  convert it to a .archimate file and open it in Archi with no more links with current repository)
* Revert a commit
* Restore an old version of the model: this revert all commit that came after the choosed version

**Comment by jbsarrodie** [comment]
> Open the version of the model at a specified commit: this would convert it to a .archimate file and open it in Archi with no more links with current repository)

As discussed in our meeting, this is exactly what I have in mind. The key point is that HEAD should not change or else next commit will be "off branch". A potential implementation would be to clone the local repository on another (temporary) folder anf checkout the commit there to load it, and them remove the folder. But this might be a bit "too much".

**Comment by Phillipus** [comment]
What is the use case for a user wanting to create a separate .archimate file from a certain commit?

**Comment by jbsarrodie** [comment]
The use case is a user wanting to review a previous state of the model (maybe to copy/paste concepts from the old version to the current one in case of unwanted changes from another user).

In the future when we l have branching we ll might want to allow to create new branch from a commit but that s another story.

**Comment by Phillipus** [comment]
> Open the version of the model at a specified commit: this would convert it to a .archimate file and open it in Archi with no more links with current repository)

First implementation has been implemented.

**Comment by Phillipus** [comment]
> Revert a commit

Do you mean a commit that has not been pushed?

> Restore an old version of the model: this revert all commit that came after the choosed version

Do you mean take the model from the commit point and re-commit this as new?

**Comment by jbsarrodie** [comment]
> Do you mean a commit that has not been pushed?

I mean create a new commit that reverts the changes done on another commit (pushed or not). This is the standard [git revert](https://git-scm.com/docs/git-revert.html).

> Do you mean take the model from the commit point and re-commit this as new? Or revert all commits that have not been pushed? Or.....?

I mean create a new commit that reverts all commit from the one just after the selected on up to the latest one. Again this is a git revert.

**Comment by Phillipus** [comment]
*Revert a commit*

Some things to be aware of:

- git revert doc says: "This requires your working tree to be clean (no modifications from the HEAD commit)"
- git revert allows more than one commit to be selected for the revert
- it will probably cause a merge conflict

**Comment by jbsarrodie** [comment]
> git revert doc says: "This requires your working tree to be clean (no modifications from the HEAD commit)"

Yes, so we must make sure user either commit or cancel his changes.

**Comment by Phillipus** [comment]
I've added an action to the History view to revert a single commit. Note:

- this is only for a single commit to keep things simple
- if merge conflict is detected the dialog will open

This is the basic implementation. We still need to resolve #7 for it to work properly.

Also, I think the user needs to know what exactly they are reverting. Can they ascertain this from the commit message? Or do they need to visualise it somehow?

**Comment by Phillipus** [comment]
> create a new commit that reverts all commit from the one just after the selected on up to the latest one. Again this is a git revert.

I've added another action to the History view to revert up to a selected commit:

There are some issues with this:

- It will fail if one of the commits has more than one parent (branch)
- If successful, the revert command will create more than one commit - one for each reverted commit. I haven't found an option to "squash" all into one commit.

**Comment by Phillipus** [comment]
I've added an "Undo Last Commit" action which is basically a "soft" reset on HEAD if the last commit has not been pushed.

It is available in the History view. I tried adding the action to the Model Repo view but it then looked cluttered with an extra action there. However, we can show the action there and/or on the models tree or main menu if required.

Assuming, that is, that this action is needed.  I found this action useful when testing commits.
